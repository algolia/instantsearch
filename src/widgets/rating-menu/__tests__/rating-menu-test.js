import { render } from 'preact-compat';
import jsHelper, {
  SearchResults,
  SearchParameters,
} from 'algoliasearch-helper';
import ratingMenu from '../rating-menu';

jest.mock('preact-compat', () => {
  const module = require.requireActual('preact-compat');

  module.render = jest.fn();

  return module;
});

describe('Usage', () => {
  it('throws without container', () => {
    expect(() => {
      ratingMenu({ container: undefined });
    }).toThrowErrorMatchingInlineSnapshot(`
"The \`container\` option is required.

See documentation: https://www.algolia.com/doc/api-reference/widgets/rating-menu/js/"
`);
  });
});

describe('ratingMenu()', () => {
  const attribute = 'anAttrName';
  let container;
  let widget;
  let helper;
  let state;
  let createURL;
  let results;

  beforeEach(() => {
    render.mockClear();

    container = document.createElement('div');
    widget = ratingMenu({
      container,
      attribute,
      cssClasses: { body: ['body', 'cx'] },
    });
    helper = jsHelper(
      {},
      '',
      widget.getWidgetSearchParameters(new SearchParameters({}), {
        uiState: {},
      })
    );
    jest.spyOn(helper, 'clearRefinements');
    jest.spyOn(helper, 'addDisjunctiveFacetRefinement');
    jest.spyOn(helper, 'getRefinements');
    jest.spyOn(helper, 'removeDisjunctiveFacetRefinement');
    helper.search = jest.fn();

    state = {
      toggleRefinement: jest.fn(),
    };
    results = {
      getFacetValues: jest.fn().mockReturnValue([]),
      hits: [],
    };
    createURL = () => '#';
    widget.init({
      helper,
      instantSearchInstance: { templatesConfig: undefined },
    });
  });

  it('configures the underlying disjunctive facet', () => {
    expect(
      widget.getWidgetSearchParameters(new SearchParameters(), { uiState: {} })
    ).toEqual(
      new SearchParameters({
        disjunctiveFacets: ['anAttrName'],
        disjunctiveFacetsRefinements: {
          anAttrName: [],
        },
      })
    );
  });

  it('calls twice render(<RefinementList props />, container)', () => {
    widget.render({ state, helper, results, createURL });
    widget.render({ state, helper, results, createURL });

    expect(render).toHaveBeenCalledTimes(2);
    expect(render.mock.calls[0][0]).toMatchSnapshot();
    expect(render.mock.calls[0][1]).toEqual(container);
    expect(render.mock.calls[1][0]).toMatchSnapshot();
    expect(render.mock.calls[1][1]).toEqual(container);
  });

  it('hide the count==0 when there is a refinement', () => {
    helper.addDisjunctiveFacetRefinement(attribute, 1);
    const _results = new SearchResults(helper.state, [
      {
        facets: {
          [attribute]: { 1: 42 },
        },
      },
      {},
    ]);

    widget.render({ state, helper, results: _results, createURL });
    expect(render).toHaveBeenCalledTimes(1);
    expect(render.mock.calls[0][0].props.facetValues).toEqual([
      {
        count: 42,
        isRefined: true,
        name: '1',
        value: '1',
        stars: [true, false, false, false, false],
      },
    ]);
  });

  it("doesn't call the refinement functions if not refined", () => {
    helper.getRefinements = jest.fn().mockReturnValue([]);
    widget.render({ state, helper, results, createURL });
    expect(helper.clearRefinements).toHaveBeenCalledTimes(0);
    expect(helper.addDisjunctiveFacetRefinement).toHaveBeenCalledTimes(0);
    expect(helper.search).toHaveBeenCalledTimes(0);
  });

  it('refines the search', () => {
    helper.getRefinements = jest.fn().mockReturnValue([]);
    widget._toggleRefinement('3');
    expect(helper.removeDisjunctiveFacetRefinement).toHaveBeenCalledTimes(1);
    expect(helper.addDisjunctiveFacetRefinement).toHaveBeenCalledTimes(3);
    expect(helper.search).toHaveBeenCalledTimes(1);
  });

  it('toggles the refinements', () => {
    helper.addDisjunctiveFacetRefinement(attribute, 2);
    helper.addDisjunctiveFacetRefinement.mockReset();
    widget._toggleRefinement('2');
    expect(helper.removeDisjunctiveFacetRefinement).toHaveBeenCalledTimes(1);
    expect(helper.addDisjunctiveFacetRefinement).toHaveBeenCalledTimes(0);
    expect(helper.search).toHaveBeenCalledTimes(1);
  });

  it('toggles the refinements with another facet', () => {
    helper.getRefinements = jest.fn().mockReturnValue([{ value: '2' }]);
    widget._toggleRefinement('4');
    expect(helper.removeDisjunctiveFacetRefinement).toHaveBeenCalledTimes(1);
    expect(helper.addDisjunctiveFacetRefinement).toHaveBeenCalledTimes(2);
    expect(helper.search).toHaveBeenCalledTimes(1);
  });

  it('should return the right facet counts and results', () => {
    const _widget = ratingMenu({
      container,
      attribute,
      cssClasses: { body: ['body', 'cx'] },
    });
    const _helper = jsHelper(
      {},
      '',
      _widget.getWidgetSearchParameters(new SearchParameters({}), {
        uiState: {},
      })
    );
    _helper.search = jest.fn();

    _widget.init({
      helper: _helper,
      state: _helper.state,
      createURL: () => '#',
      instantSearchInstance: {
        templatesConfig: {},
      },
    });

    _widget.render({
      results: new SearchResults(_helper.state, [
        {
          facets: {
            [attribute]: { 0: 5, 1: 10, 2: 20, 3: 50, 4: 900, 5: 100 },
          },
        },
        {},
      ]),
      state: _helper.state,
      helper: _helper,
      createURL: () => '#',
      instantSearchInstance: {
        templatesConfig: {},
      },
    });

    expect(
      render.mock.calls[render.mock.calls.length - 1][0].props.facetValues
    ).toEqual([
      {
        count: 1000,
        isRefined: false,

        name: '4',
        value: '4',
        stars: [true, true, true, true, false],
      },
      {
        count: 1050,
        isRefined: false,

        name: '3',
        value: '3',
        stars: [true, true, true, false, false],
      },
      {
        count: 1070,
        isRefined: false,

        name: '2',
        value: '2',
        stars: [true, true, false, false, false],
      },
      {
        count: 1080,
        isRefined: false,

        name: '1',
        value: '1',
        stars: [true, false, false, false, false],
      },
    ]);
  });
});
