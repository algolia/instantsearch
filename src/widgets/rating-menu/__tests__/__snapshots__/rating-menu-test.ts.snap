// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ratingMenu() calls twice render(<RefinementList props />, container) 1`] = `
{
  "createURL": [Function],
  "cssClasses": {
    "count": "ais-RatingMenu-count",
    "disabledItem": "ais-RatingMenu-item--disabled",
    "emptyStarIcon": "ais-RatingMenu-starIcon--empty",
    "fullStarIcon": "ais-RatingMenu-starIcon--full",
    "item": "ais-RatingMenu-item",
    "label": "ais-RatingMenu-label",
    "link": "ais-RatingMenu-link",
    "list": "ais-RatingMenu-list",
    "noRefinementRoot": "ais-RatingMenu--noRefinement",
    "root": "ais-RatingMenu",
    "selectedItem": "ais-RatingMenu-item--selected",
    "starIcon": "ais-RatingMenu-starIcon",
  },
  "depth": 0,
  "facetValues": [
    {
      "count": 0,
      "isRefined": false,
      "label": "4",
      "name": "4",
      "stars": [
        true,
        true,
        true,
        true,
        false,
      ],
      "value": "4",
    },
    {
      "count": 0,
      "isRefined": false,
      "label": "3",
      "name": "3",
      "stars": [
        true,
        true,
        true,
        false,
        false,
      ],
      "value": "3",
    },
    {
      "count": 0,
      "isRefined": false,
      "label": "2",
      "name": "2",
      "stars": [
        true,
        true,
        false,
        false,
        false,
      ],
      "value": "2",
    },
    {
      "count": 0,
      "isRefined": false,
      "label": "1",
      "name": "1",
      "stars": [
        true,
        false,
        false,
        false,
        false,
      ],
      "value": "1",
    },
  ],
  "templateProps": {
    "templates": {
      "item": "{{#count}}<a class=\\"{{cssClasses.link}}\\" aria-label=\\"{{value}} & up\\" href=\\"{{url}}\\">{{/count}}{{^count}}<div class=\\"{{cssClasses.link}}\\" aria-label=\\"{{value}} & up\\" disabled>{{/count}}
  {{#stars}}<svg class=\\"{{cssClasses.starIcon}} {{#.}}{{cssClasses.fullStarIcon}}{{/.}}{{^.}}{{cssClasses.emptyStarIcon}}{{/.}}\\" aria-hidden=\\"true\\" width=\\"24\\" height=\\"24\\">
    {{#.}}<use xlink:href=\\"#ais-RatingMenu-starSymbol\\"></use>{{/.}}{{^.}}<use xlink:href=\\"#ais-RatingMenu-starEmptySymbol\\"></use>{{/.}}
  </svg>{{/stars}}
  <span class=\\"{{cssClasses.label}}\\">& Up</span>
  {{#count}}<span class=\\"{{cssClasses.count}}\\">{{#helpers.formatNumber}}{{count}}{{/helpers.formatNumber}}</span>{{/count}}
{{#count}}</a>{{/count}}{{^count}}</div>{{/count}}",
    },
    "templatesConfig": undefined,
    "useCustomCompileOptions": {
      "item": false,
    },
  },
  "toggleRefinement": [Function],
}
`;
